src/base/http/requestparser.cpp

```diff
@@ -30,314 +31,307 @@
+            if (!part.isEmpty() || (behavior == QString::KeepEmptyParts))
+                ret += part;
+
+            head = end + sep.size();
+        }
+
+        return ret;
+    }
 }
 
-RequestParser::RequestParser(uint maxContentLength)
-    : m_maxContentLength(maxContentLength)
+RequestParser::RequestParser()
+    : m_request()
 {
 }
 
-RequestParser::ErrorCode RequestParser::parseHttpRequest(const QByteArray &data, Request &request)
+RequestParser::ParseResult RequestParser::parse(const QByteArray &data)
 {
-    m_request = Request();
+    // Warning! Header names are converted to lowercase
+    return RequestParser().doParse(data);
+}
 
-    // Parse HTTP request header
+RequestParser::ParseResult RequestParser::doParse(const QByteArray &data)
+{
+    // for simplicity, we don't handle malformed requests which use double `LF` as delimiter
     const int headerEnd = data.indexOf(EOH);
     if (headerEnd < 0) {
         qDebug() << Q_FUNC_INFO << "incomplete request";
-        return IncompleteRequest;
+        return {ParseStatus::Incomplete, Request(), 0};
     }
 
-    if (!parseHttpHeader(data.left(headerEnd))) {
+    const QString httpHeaders = QString::fromLatin1(data.left(headerEnd));
+    if (!parseStartLines(httpHeaders)) {
         qWarning() << Q_FUNC_INFO << "header parsing error";
-        return BadRequest;
+        return {ParseStatus::BadRequest, Request(), 0};
     }
 
-    // Parse HTTP request message
-    if (m_request.headers.contains("content-length")) {
-        int contentLength = m_request.headers["content-length"].toInt();
-        if (contentLength < 0) {
-            qWarning() << Q_FUNC_INFO << "bad request: content-length is negative";
-            return BadRequest;
-        }
+    const int headerLength = headerEnd + EOH.length();
 
-        if (contentLength > static_cast<int>(m_maxContentLength)) {
+    // handle supported methods
+    if ((m_request.method == HEADER_REQUEST_METHOD_GET) || (m_request.method == HEADER_REQUEST_METHOD_HEAD)) {
+        return {ParseStatus::OK, m_request, headerLength};
+    }
+    else if (m_request.method == HEADER_REQUEST_METHOD_POST) {
+        bool ok = false;
+        const int contentLength = m_request.headers[HEADER_CONTENT_LENGTH].toInt(&ok);
+        if (!ok || (contentLength < 0)) {
+            qWarning() << Q_FUNC_INFO << "bad request: content-length invalid";
+            return {ParseStatus::BadRequest, Request(), 0};
+        }
+        if (contentLength > MAX_REQUEST_SIZE) {
             qWarning() << Q_FUNC_INFO << "bad request: message too long";
-            return BadRequest;
+            return {ParseStatus::BadRequest, Request(), 0};
         }
 
-        QByteArray content = data.mid(headerEnd + EOH.length(), contentLength);
-        if (content.length() < contentLength) {
-            qDebug() << Q_FUNC_INFO << "incomplete request";
-            return IncompleteRequest;
-        }
+        if (contentLength > 0) {
```

Comment:
This check is redundant. It hides some kind of bad request. If content length header is 0, then only requests with empty body is valid, isn't it?

Reply:
>If content length header is 0, then only requests with empty body is valid, isn't it?

Good catch! I'll update this part.

Reply:
>If content length header is 0, then only requests with empty body is valid, isn't it?

Now I read again, the situation is, with http persistence connection, a series of requests could be concatenated & received together.
Currently I choose the easy way to parse: if the packet is malformed (`contentLength == 0` + body), parsing will succeed for the 1st round but will fail at the 2nd round (now the packet is malformed the connection will be closed).

>This check is redundant. 

Given the above, this check is actually skipping over unneeded processing when possible.